internal/graph/checkingresourcestream.go

```diff
@@ -0,0 +1,537 @@
+	if len(toProcess) == 0 {
+		return false, nil
+	}
+
+	for _, current := range toProcess {
+		if current.reachableResult.Resource.ResultStatus == v1.ReachableResource_HAS_PERMISSION {
+			current.lookupResult = &v1.DispatchLookupResourcesResponse{
+				ResolvedResource: &v1.ResolvedResource{
+					ResourceId:     current.reachableResult.Resource.ResourceId,
+					Permissionship: v1.ResolvedResource_HAS_PERMISSION,
+				},
+				Metadata:            addCallToResponseMetadata(current.reachableResult.Metadata),
+				AfterResponseCursor: current.reachableResult.AfterResponseCursor,
+			}
+			crs.rq.updateToBePublished(current)
+			continue
+		}
+
+		toCheck.Add(current.reachableResult.Resource.ResourceId, current)
+	}
+
+	if toCheck.IsEmpty() {
+		crs.availableForPublishing <- true
+		return true, nil
+	}
+
+	// Issue the bulk check over all the resources.
+	results, resultsMetas, err := computed.ComputeBulkCheck(
+		crs.ctx,
+		crs.checker,
+		computed.CheckParameters{
+			ResourceType:  crs.req.ObjectRelation,
+			Subject:       crs.req.Subject,
+			CaveatContext: crs.req.Context.AsMap(),
+			AtRevision:    crs.req.Revision,
+			MaximumDepth:  crs.req.Metadata.DepthRemaining,
+			DebugOption:   computed.NoDebugging,
+		},
+		toCheck.Keys(),
+	)
+	if err != nil {
+		return true, err
+	}
+
+	for _, rai := range toCheck.Values() {
+		checkResult := results[rai.reachableResult.Resource.ResourceId]
+
+		var permissionship v1.ResolvedResource_Permissionship
+		var missingFields []string
+
+		switch {
+		case checkResult == nil || checkResult.Membership == v1.ResourceCheckResult_NOT_MEMBER:
+			permissionship = v1.ResolvedResource_UNKNOWN
+
+		case checkResult != nil && checkResult.Membership == v1.ResourceCheckResult_MEMBER:
+			permissionship = v1.ResolvedResource_HAS_PERMISSION
+
+		case checkResult != nil && checkResult.Membership == v1.ResourceCheckResult_CAVEATED_MEMBER:
+			permissionship = v1.ResolvedResource_CONDITIONALLY_HAS_PERMISSION
+			missingFields = checkResult.MissingExprFields
+
+		default:
+			return true, spiceerrors.MustBugf("unknown check result status for reachable resources")
+		}
+
+		// Set the lookupResult iff the permissionship was a valid permission.
+		var lookupResult *v1.DispatchLookupResourcesResponse
+		if permissionship != v1.ResolvedResource_UNKNOWN {
+			metadata := combineResponseMetadata(rai.reachableResult.Metadata, resultsMetas[rai.reachableResult.Resource.ResourceId])
+			lookupResult = &v1.DispatchLookupResourcesResponse{
+				ResolvedResource: &v1.ResolvedResource{
+					ResourceId:             rai.reachableResult.Resource.ResourceId,
+					Permissionship:         permissionship,
+					MissingRequiredContext: missingFields,
+				},
+				Metadata:            metadata,
+				AfterResponseCursor: rai.reachableResult.AfterResponseCursor,
+			}
+		}
+
+		rai.lookupResult = lookupResult
+		crs.rq.updateToBePublished(rai)
+	}
+
+	crs.availableForPublishing <- true
```

Comment:
wouldn't we want to signal element is available for publishing as soon as it is published, or you think that'd be much signaling overhead, and have `publishResourcesIfPossible` invoked on batches rather than on every element that becomes ready?

Reply:
I figured the batch would make things slightly more efficient